From 41cd3370adc24bf41923909e9edb3ddbe11adff1 Mon Sep 17 00:00:00 2001
From: feeRnt <81442162+feeRnt@users.noreply.github.com>
Date: Sun, 24 Aug 2025 20:47:51 +0000
Subject: [PATCH] Merge commit from planned ps4-linux-6.15.y branch.

mmc: Fix SDIO WiFi + Bluetooth for Marvell 88w8897

Fixes card detection and functionality for the Marvell 88w8897 Chip
(PlayStation internal codename Torus 2.0).

The affected cards are present in the console models :-
CUH-1216{A,B},
CUH-1215{A,B},
CUH-1003{A},
... and maybe more.

Files changed:
- include/linux/mmc/sdio_ids.h:
Marvell 8897 F0 (function 0) ID in include/linux/mmc/sdio_ids.h.
To be polled by SDIO/SDHCI driver during card func initialization.

- drivers/mmc/core/quirks.h
Added MMC SDR104 mode with 150 MHz quirk for MARVELL_8897_F0 in drivers/mmc/core/quirks.h.

- drivers/mmc/host/sdhci-pci-core.c
Added SDHCI host controller quirk for correct clock tuning on affected
systems.
Adding .quirks2 = SDHCI_QUIRK2_PRESET_VALUE_BROKEN,
in drivers/mmc/host/sdhci-pci-core.c .

Added a check for SD host controller version (Aeolia or Belize only) before applying
the host quirk.

- drivers/mmc/core/sdio_ops.c
Added skeleton comment to have 3 retries on each CMD52 or 53 in mmc_wait_for_cmd function.
Is not needed now, but for future reference in case some card fails at
card function init.

Signed-off-by: feeRnt <81442162+feeRnt@users.noreply.github.com>
---
 drivers/mmc/core/quirks.h         |  2 +-
 drivers/mmc/core/sdio_ops.c       | 10 +++++++---
 drivers/mmc/host/sdhci-pci-core.c | 26 ++++++++++++++++++++++++++
 include/linux/mmc/sdio_ids.h      |  2 +-
 4 files changed, 35 insertions(+), 5 deletions(-)

diff --git a/drivers/mmc/core/quirks.h b/drivers/mmc/core/quirks.h
index f8a94dbe067a934ba815526c6fd0f855b79861d3..bb8902f206a4e132e70daa41d1e259609c65cc38 100644
--- a/drivers/mmc/core/quirks.h
+++ b/drivers/mmc/core/quirks.h
@@ -209,7 +209,7 @@ static const struct mmc_fixup __maybe_unused sdio_fixup_methods[] = {
 		   add_limit_rate_quirk, 150000000),
 	/* On multi function cards, the base, non wifi, non BT, function, is called F0
 	 * Since 8897 traditionally doesn't have any need for any quirk handling, the SDIO_IDs
-	 * file doesn't include the F0 ID. which is the 0x912c. (I assume).
+	 * file doesn't include the F0 ID. which is the 0x912c.
 	 */
 	SDIO_FIXUP(SDIO_VENDOR_ID_MARVELL, SDIO_DEVICE_ID_MARVELL_8897_F0,
 		   add_limit_rate_quirk, 150000000),
diff --git a/drivers/mmc/core/sdio_ops.c b/drivers/mmc/core/sdio_ops.c
index 955f03214646f641b42aa2b8ba3f29cf1a5df969..647fd66444770c03753c87ce581d605de1a9c435 100644
--- a/drivers/mmc/core/sdio_ops.c
+++ b/drivers/mmc/core/sdio_ops.c
@@ -80,11 +80,15 @@ static int mmc_io_rw_direct_host(struct mmc_host *host, int write, unsigned fn,
 	cmd.arg |= in;
 	cmd.flags = MMC_RSP_SPI_R5 | MMC_RSP_R5 | MMC_CMD_AC;
 
-	/* err = mmc_wait_for_cmd(host, &cmd, 0);
-	 * https://fa.linux.kernel.narkive.com/t6hn0noQ/patch-mmc-sdio-retry-cmd52-53-when-error-happens
+	err = mmc_wait_for_cmd(host, &cmd, 0);
+	/* https://fa.linux.kernel.narkive.com/t6hn0noQ/patch-mmc-sdio-retry-cmd52-53-when-error-happens
 	 * Use 3 retries instead of 0 retries:
+	 *
+	 * err = mmc_wait_for_cmd(host, &cmd, MMC_CMD_RETRIES);
+	 * Although proposed by a Marvell engineer, it now seems unnecessary with
+	 * the current patches - on the PS4 88w8897 cards with errors at sdio_init_func.
+	 * Keeping the comment for future reference if some card fails to initialize.
 	 */
-	err = mmc_wait_for_cmd(host, &cmd, MMC_CMD_RETRIES);
 	if (err)
 		return err;
 
diff --git a/drivers/mmc/host/sdhci-pci-core.c b/drivers/mmc/host/sdhci-pci-core.c
index e8632b203ab06380f76aceac82c093610c9f7c2a..d4c44275ec4c88631934e2ebee2173a8e047b248 100644
--- a/drivers/mmc/host/sdhci-pci-core.c
+++ b/drivers/mmc/host/sdhci-pci-core.c
@@ -350,6 +350,32 @@ static int aeolia_probe_slot(struct sdhci_pci_slot *slot)
 		return -ENODEV;
 	}
 	slot->host->irq = slot->chip->pdev->irq;
+
+	/**			-- Quirk Handling --			**
+	* Needed by certain PS4 console models with the Marvell 88w8897 WiFi+BT chip,
+	* typically present on Belize CUH-1215 and CUH-1216 motherboards.
+	*
+	* Although there's a pathway to identify the model using the AMD VGA controller ID:
+	* #define PCI_DEVICE_ID_CUH_11XX 0x9920
+	* #define PCI_DEVICE_ID_CUH_12XX 0x9922
+	* from drivers/gpu/drm/amd/amdgpu/ps4_bridge.c,
+	*
+	* there is report from an Aeolia CUH-1003 model as well, with the same issue.
+
+	* So it's best to use the SDHCI IDs directly.
+	* Ideally we'd match against the card's CIS vendor and device IDs as well,
+	* but since this host controller has to initialize before we init the card,
+	* there seems to be no good way to check that.
+	*
+	* On unaffected systems, the quirk only prevents usage of predefined clock timings,
+	* for standard clock modes (SDR50, SDR104), and instead calculates it on the fly
+	* using host capabiilites. So it should be harmless in all cases.
+	*/
+	if ((slot->chip->pdev->device == PCI_DEVICE_ID_SONY_AEOLIA_SDHCI) || \
+	    (slot->chip->pdev->device == PCI_DEVICE_ID_SONY_BELIZE_SDHCI)) {
+		slot->host->quirks2 |= SDHCI_QUIRK2_PRESET_VALUE_BROKEN;
+        }
+
 	return 0;
 }
 
diff --git a/include/linux/mmc/sdio_ids.h b/include/linux/mmc/sdio_ids.h
index 160e4a6923a3de254c19bc8f9a7f9d3e46c22be8..eb0114991077214db01bcb682ab61075381b402c 100644
--- a/include/linux/mmc/sdio_ids.h
+++ b/include/linux/mmc/sdio_ids.h
@@ -93,7 +93,7 @@
 #define SDIO_DEVICE_ID_MARVELL_8797_F0		0x9128
 #define SDIO_DEVICE_ID_MARVELL_8797_WLAN	0x9129
 #define SDIO_DEVICE_ID_MARVELL_8797_BT		0x912a
-#define SDIO_DEVICE_ID_MARVELL_8897_F0		0x912c //Marvell_8897_F0 PS4_1216
+#define SDIO_DEVICE_ID_MARVELL_8897_F0		0x912c //Marvell_8897_F0 - In PS4_12xx
 #define SDIO_DEVICE_ID_MARVELL_8897_WLAN	0x912d
 #define SDIO_DEVICE_ID_MARVELL_8897_BT		0x912e
 #define SDIO_DEVICE_ID_MARVELL_8887_F0		0x9134
